src/v/storage/record_batch_builder.cc

```diff
@@ -45,6 +47,14 @@ model::record_batch record_batch_builder::build() && {
       .ctx = model::record_batch_header::context(
         model::term_id(0), ss::this_shard_id())};
 
+    /**
+     * for all batches other than data batches set control batch type attribute,
+     * this way clients will be able to skip those batches while fetching
+     */
+    if (_batch_type != data_batch_type) {
+        header.attrs.set_control_type();
+    }
```

Comment:
should we _only_ set the control bit and do the record manipulation on non-data record batches in fetch.cc?

1. that would seem to be sufficient to address all of the issues this PR touches
2. none of our internal batch handling requires this bit
3. it doesn't require us to know that this is exactly how we want to use this bit in the future (e.g. waiting for transactions)

Reply:
Good point, I will drop this commit